File: datasets/lua/HumanEval_1_separate_paren_groups.lua
-- Input to this function is a string containing multiple groups of nested parentheses. Your goal is to
-- separate those group into separate strings and return the list of those.
-- Separate groups are balanced (each open brace is properly closed) and not nested within each other
-- Ignore any spaces in the input string.
-- >>> separate_paren_groups('( ) (( )) (( )( ))')
-- ['()', '(())', '(()())']
local function separate_paren_groups(paren_string)
  local paren_groups = {}
  local paren_group = ""
  local paren_count = 0
  for i = 1, #paren_string do
    local char = paren_string:sub(i, i)
    if char == "(" then
      paren_count = paren_count + 1
      paren_group = paren_group .. char
    elseif char == ")" then
      paren_count = paren_count - 1
      paren_group = paren_group .. char
    elseif char == " " then
      if paren_count == 0 then
        table.insert(paren_groups, paren_group)
        paren_group = ""
      end
    else
      paren_group = paren_group .. char
    end
  end
  if paren_group ~= "" then
    table.insert(paren_groups, paren_group)
  end
  return paren_groups
end

lu = require('luaunit')

function test_humaneval()
local candidate = separate_paren_groups
    lu.assertEquals(candidate('(()()) ((())) () ((())()())'), {'(()())', '((()))', '()', '((())()())'})
    lu.assertEquals(candidate('() (()) ((())) (((())))'), {'()', '(())', '((()))', '(((())))'})
    lu.assertEquals(candidate('(()(())((())))'), {'(()(())((())))'})
    lu.assertEquals(candidate('( ) (( )) (( )( ))'), {'()', '(())', '(()())'})
end

os.exit(lu.LuaUnit.run())

File: datasets/java/HumanEval_1_separate_paren_groups.java
import java.util.*;
import java.lang.reflect.*;
import org.javatuples.*;
class Problem {
private static <T> T getDefaultValue(Class<T> clazz) {
return (T) Array.get(Array.newInstance(clazz, 1), 0);
}    // Input to this function is a string containing multiple groups of nested parentheses. Your goal is to
    // separate those group into separate strings and return the list of those.
    // Separate groups are balanced (each open brace is properly closed) and not nested within each other
    // Ignore any spaces in the input string.
    // >>> separate_paren_groups('( ) (( )) (( )( ))')
    // ['()', '(())', '(()())']
    public static ArrayList<String> separate_paren_groups(String paren_string) {
        ArrayList<String> result = new ArrayList<>();
        Stack<Character> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();
        for (char c : paren_string.toCharArray()) {
            if (c == '(') {
                stack.push(c);
                sb.append(c);
            } else if (c == ')') {
                if (stack.isEmpty()) {
                    sb.append(c);
                } else {
                    stack.pop();
                    if (stack.isEmpty()) {
                        result.add(sb.toString());
                        sb = new StringBuilder();
                    }
                }
            }
        }
        return result;
    }
    // Input to this function is a string containing multiple groups of nested parentheses. Your goal is to
    // separate those group into separate strings and return the list of those.
    // Separate groups are balanced (each open brace is properly closed) and not nested within each other
    // Ignore any spaces in the input string.
    // >>> separate_paren_groups('( ) (( )) (( )( ))')
    // ['()', '(())', '(()())']
    public static ArrayList<String> separate_paren_groups(String paren_string) {
        ArrayList<String> result = new ArrayList<>();
        Stack<Character> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();
        for (char c : paren_string.toCharArray()) {
            if (c == '(') {
                stack.push(c);
                sb.append(c);
            } else if (c == ')') {
                if (stack.isEmpty()) {
                    sb.append(c);
                } else {
                    stack.pop();
                    if (stack.isEmpty()) {
                        result.add(sb.toString());
                        sb = new StringBuilder();
                    }
                }
            }
        }
        return result;
    }
    // Input to this function is a string containing multiple groups of nested parentheses. Your goal is to
    // separate those

    }
    public static void main(String[] args) {
    separate_paren_groups(("(()()) ((())) () ((())()())")).equals((new ArrayList<String>(Arrays.asList((String)"(()())", (String)"((()))", (String)"()", (String)"((())()())"))));
    separate_paren_groups(("() (()) ((())) (((())))")).equals((new ArrayList<String>(Arrays.asList((String)"()", (String)"(())", (String)"((()))", (String)"(((())))"))));
    separate_paren_groups(("(()(())((())))")).equals((new ArrayList<String>(Arrays.asList((String)"(()(())((())))"))));
    separate_paren_groups(("( ) (( )) (( )( ))")).equals((new ArrayList<String>(Arrays.asList((String)"()", (String)"(())", (String)"(()())"))));
    }

}
